sdk/messaging/azservicebus/client.go

```diff
@@ -111,72 +116,123 @@ func newClientImpl(config clientConfig, options ...ClientOption) (*Client, error
+	id, cleanupOnClose := client.getCleanupForCloseable()
+
+	processor, err := newProcessor(client.namespace, cleanupOnClose, options...)
 
 	if err != nil {
 		return nil, err
 	}
 
-	// TODO: clean up these links
-	client.linksMu.Lock()
-	client.links = append(client.links, sender)
-	client.linksMu.Unlock()
-
-	return sender, nil
+	client.addCloseable(id, processor)
+	return processor, nil
 }
 
-// NewReceiver creates a Receiver, which allows you to receive messages.
-func (client *Client) NewReceiver(options ...ReceiverOption) (*Receiver, error) {
-	receiver, err := newReceiver(client.namespace, options...)
+// NewReceiver creates a Receiver for a queue. A receiver allows you to receive messages.
+func (client *Client) NewReceiverForQueue(queue string, options ...ReceiverOption) (*Receiver, error) {
+	id, cleanupOnClose := client.getCleanupForCloseable()
+	options = append(options, receiverWithQueue(queue))
+
+	receiver, err := newReceiver(client.namespace, cleanupOnClose, options...)
 
 	if err != nil {
 		return nil, err
 	}
 
-	// TODO: clean up these links
-	client.linksMu.Lock()
-	client.links = append(client.links, receiver)
-	client.linksMu.Unlock()
+	client.addCloseable(id, receiver)
+	return receiver, nil
+}
+
+// NewReceiver creates a Receiver for a subscription. A receiver allows you to receive messages.
+func (client *Client) NewReceiverForSubscription(topic string, subscription string, options ...ReceiverOption) (*Receiver, error) {
+	id, cleanupOnClose := client.getCleanupForCloseable()
+	options = append(options, receiverWithSubscription(topic, subscription))
+
+	receiver, err := newReceiver(client.namespace, cleanupOnClose, options...)
 
+	if err != nil {
+		return nil, err
+	}
+
+	client.addCloseable(id, receiver)
 	return receiver, nil
 }
 
+// NewSender creates a Sender, which allows you to send messages or schedule messages.
+func (client *Client) NewSender(queueOrTopic string) (*Sender, error) {
+	id, cleanupOnClose := client.getCleanupForCloseable()
+	sender, err := newSender(client.namespace, queueOrTopic, cleanupOnClose)
+
+	if err != nil {
+		return nil, err
+	}
+
+	client.addCloseable(id, sender)
+	return sender, nil
+}
+
 // Close closes the current connection Service Bus as well as any Sender, Receiver or Processors created
 // using this client.
 func (client *Client) Close(ctx context.Context) error {
 	var lastError error
 
+	var links []interface {
+		Close(ctx context.Context) error
+	}
+
 	client.linksMu.Lock()
 
 	for _, link := range client.links {
+		links = append(links, link)
+	}
+
+	client.linksMu.Unlock()
+
+	for _, link := range links {
 		if err := link.Close(ctx); err != nil {
 			tab.For(ctx).Error(err)
 			lastError = err
 		}
 	}
 
-	client.linksMu.Unlock()
-
 	if lastError != nil {
 		return fmt.Errorf("errors while closing links: %w", lastError)
 	}
 	return nil
 }
+
+func (client *Client) addCloseable(id uint64, closeable interface {
+	Close(ctx context.Context) error
+}) {
```

Comment:
nit: You use the `closeable` interface you have defined here in other spots. Why not define this interface as a `closeable` to reduce code?

Reply:
Done.